File: js/algorithms/virtual_memory/thrashing.js
/*
 * thrashing.js
 * Models thrashing by simulating N processes with working sets that may
 * exceed their allocated frames. As more processes are added, CPU
 * utilisation collapses as all time is spent on page faults.
 * JS port of c/algorithms/virtual_memory/thrashing.c
 */

export function thrashingSimulate(input) {
  const { processCount, framesPerProcess, workingSetSizes, totalFrames } = input;
  const steps = [];
  let thrashingOnsetStep = -1;
  let peakFaultRate = 0;

  for (let step = 1; step <= processCount; step++) {
    let totalWs = 0, totalFaults = 0;
    for (let p = 0; p < step; p++) {
      totalWs += workingSetSizes[p];
      const frames = framesPerProcess[p];
      const ws = workingSetSizes[p];
      if (ws > frames) totalFaults += (ws - frames) * 2;
    }

    let cpuUtil, thrashing;
    if (totalWs <= totalFrames) {
      cpuUtil = Math.min(95, 85 + step * 2);
      thrashing = false;
    } else {
      const excess = totalWs - totalFrames;
      const drop = Math.round((excess * 80) / (totalFrames || 1));
      cpuUtil = Math.max(5, 95 - drop);
      thrashing = cpuUtil < 30;
      if (thrashing && thrashingOnsetStep === -1) thrashingOnsetStep = step;
    }

    if (totalFaults > peakFaultRate) peakFaultRate = totalFaults;

    steps.push({
      timeStep: step,
      cpuUtilization: cpuUtil,
      thrashing,
      activeProcesses: step,
      totalPageFaults: totalFaults
    });
  }

  return {
    steps,
    stepCount: steps.length,
    thrashingOnsetStep,
    peakFaultRate
  };
}

export function thrashingDetect(result) {
  return result.thrashingOnsetStep;
}

File: c/algorithms/virtual_memory/thrashing.c
/*
 * thrashing.c
 * Models thrashing by simulating N processes each with a working set
 * larger than their allocated frames. As more processes are added,
 * the sum of working sets exceeds total frames, causing CPU utilisation
 * to collapse as all time is spent handling page faults.
 * Reports CPU utilisation and fault rate at each time step.
 * Time complexity: O(process_count^2).
 */

#include "thrashing.h"
#include <string.h>

ThrashingResult thrashing_simulate(ThrashingInput input) {
    ThrashingResult result;
    memset(&result, 0, sizeof(ThrashingResult));
    result.thrashing_onset_step = -1;

    int step, p;
    for (step = 1; step <= input.process_count && step <= 256; step++) {
        ThrashingStep *s = &result.steps[result.step_count];
        s->time_step = step;
        s->active_processes = step;

        /* Sum of working set sizes for active processes */
        int total_ws = 0;
        int total_faults = 0;
        for (p = 0; p < step; p++) {
            total_ws += input.working_set_sizes[p];
            int frames = input.frames_per_process[p];
            int ws = input.working_set_sizes[p];
            if (ws > frames) total_faults += (ws - frames) * 2;
        }

        /* CPU utilisation model:
         * If sum(WS) <= total_frames: high utilisation
         * Otherwise drops proportionally */
        if (total_ws <= input.total_frames) {
            s->cpu_utilization = 85 + (step * 2);
            if (s->cpu_utilization > 95) s->cpu_utilization = 95;
            s->thrashing = 0;
        } else {
            int excess = total_ws - input.total_frames;
            int drop = (excess * 80) / (input.total_frames > 0 ? input.total_frames : 1);
            s->cpu_utilization = 95 - drop;
            if (s->cpu_utilization < 5) s->cpu_utilization = 5;
            s->thrashing = (s->cpu_utilization < 30) ? 1 : 0;
            if (s->thrashing && result.thrashing_onset_step == -1)
                result.thrashing_onset_step = step;
        }

        s->total_page_faults = total_faults;
        if (total_faults > result.peak_fault_rate)
            result.peak_fault_rate = total_faults;

        result.step_count++;
    }
    return result;
}

int thrashing_detect(ThrashingResult *result) {
    return result->thrashing_onset_step;
}
